src/Adapter/MSTest.TestAdapter/Execution/TestExecutionManager.cs

```diff
@@ -43,34 +43,35 @@ internal TestExecutionManager(IEnvironment environment, Func<Action, Task>? task
         TestMethodFilter = new TestMethodFilter();
         _sessionParameters = new Dictionary<string, object>();
         _environment = environment;
-        _taskFactory = taskFactory ?? ThreadingHelper.RunAsync;
+        _taskFactory = taskFactory ?? DefaultFactoryAsync;
     }
 
-    internal static class ThreadingHelper
+    private static Task DefaultFactoryAsync(Action action)
     {
-        public static Task RunAsync(Action action)
+        if (MSTestSettings.RunConfigurationSettings.ExecutionApartmentState == ApartmentState.STA
+            && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
         {
-            if (MSTestSettings.RunConfigurationSettings.ExecutionApartmentState == ApartmentState.STA
-                && RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            TaskCompletionSource<int> tcs = new();
+            Thread entryPointThread = new(() =>
             {
-                Thread entryPointThread = new(new ThreadStart(action));
-
-                entryPointThread.SetApartmentState(ApartmentState.STA);
-                entryPointThread.Start();
-                return Task.Factory.StartNew(
-                    entryPointThread.Join,
-                    CancellationToken.None,
-                    TaskCreationOptions.LongRunning,
-                    TaskScheduler.Default);
-            }
-            else
-            {
-                return Task.Factory.StartNew(
-                    action,
-                    CancellationToken.None,
-                    TaskCreationOptions.LongRunning,
-                    TaskScheduler.Default);
-            }
+                try
+                {
+                    action();
+                    tcs.SetResult(0);
+                }
+                catch (Exception ex)
+                {
+                    tcs.SetException(ex);
+                }
+            });
+
+            entryPointThread.SetApartmentState(ApartmentState.STA);
+            entryPointThread.Start();
+            return tcs.Task;
+        }
+        else
+        {
+            return Task.Run(action);
```

Comment:
Depending on the assumptions of the code base this change might have undesirable consequences. If the action delegate can execute arbitrary long running work and has to be assumed to do so the previous way was superior

See https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#avoid-using-taskrun-for-long-running-work-that-blocks-the-thread

Reply:
The rationale is that we expect usually quick tests like unit tests with a lot of data sets. In that case we save the creation of custom expensive threads (long running will create new thread, we see it often in the profilers and we wait the creation of all before to start the tests, so for few quick tests we're paying for no reason).

If tests are "slow" anyway we/you will have to wait...so follow the thread pool injection policy that is handling also the blocking api won't make a lot of difference compared to custom threads.

Reply:
Thanks for the explanation. Sorry for the drive by comments. I read through the change list and saw this change. Got curious about the reasons behind the nature of the change and since the PR didn't contain much more information I figured I aks.

Reply:
Some more thoughts on it, here we're not in an asp.net application where you should free the thread pool soon as possible to have a good throughput. There you don't want that your "background" workers steal resources and decrease the RPS, the background work can be done in async, handle a request needs to be done soon as possible. 

In testing world/engine, tests are our "requests" so the thread pool is our "scheduler", the thread pool solves a lot of issues related to how to scale w.t.r. the amount of available resources.

Reply:
> Thanks for the explanation. Sorry for the drive by comments. I read through the change list and saw this change. Got curious about the reasons behind the nature of the change and since the PR didn't contain much more information I figured I ask.

No worries and my bad, I did the change after a quick internal sync and completely forget to properly describe.

Reply:
FYI I linked to those guidelines not because I wanted to necessarily compare to asp.net. Many of those guidelines apply to broader scenarios and it was the quickest available source to link while I was typing my thoughts.

What I see on CI/CD environments is that they often run on shared runners with limited CPU resources available. But given the added explanation I see the tradeoffs make sense

Reply:
> What I see on CI/CD environments is that they often run on shared runners with limited CPU resources available. But given the added explanation I see the tradeoffs make sense

True, anyway we're limiting the number of workers by default to the numbers of CPU and you can decide the degree with https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.testtools.unittesting.parallelizeattribute